### PR TITLE
Azconfig aad1

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py [hunk 1]
```diff
@@ -107,8 +123,37 @@ def _create_appconfig_pipeline(self, **kwargs):
...
+
+    def _create_appconfig_aad_pipeline(self, base_url, credential, **kwargs):
+        transport = kwargs.get('transport')
+        policies = kwargs.get('policies')
+        if base_url.endswith('/'):
+            scope = base_url + '.default'
+        else:
+            scope = base_url + '/.default'
+
+        if policies is None:  # [] is a valid policy list
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py]: Not a huge fan of duplicating the list of policies twice, with only one diff on the credential policy.
If tomorrow we add a new type of policy for some reason, the likeliness of forgetting to update one of the two lists is real.
Also, we should take advantage of the fact that the pipeline skip policies that are None.

I'd rather have one method, with some code like:
```python
policies = [
  self._config.headers_policy,
  self._config.user_agent_policy,
  AppConfigRequestsCredentialsPolicy(self._config.credentials) if aad_mode else None
  self._config.retry_policy,
  SyncTokenPolicy(),
  self._config.logging_policy,  # HTTP request/response log
  DistributedTracingPolicy(**kwargs),
  HttpLoggingPolicy(**kwargs)
]
```

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py [hunk 2]
```diff
@@ -113,8 +129,40 @@ def _create_appconfig_pipeline(self, **kwargs):
...
+
+    def _create_appconfig_aad_pipeline(self, base_url, credential, **kwargs):
+        transport = kwargs.get('transport')
+        policies = kwargs.get('policies')
+        if base_url.endswith('/'):
+            scope = base_url + '.default'
+        else:
+            scope = base_url + '/.default'
+
+        if policies is None:  # [] is a valid policy list
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py]: Same comment as sync

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py [hunk 3]
```diff
@@ -113,8 +129,40 @@ def _create_appconfig_pipeline(self, **kwargs):
...
+            if hasattr(credential, "get_token"):
+                credential_policy = AsyncBearerTokenCredentialPolicy(credential, scope)
+            else:
+                raise TypeError("Unsupported credential: {}".format(credential))
+            policies = [
+                self._config.headers_policy,
+                self._config.user_agent_policy,
                 self._config.retry_policy,
+                credential_policy,
+                SyncTokenPolicy(),
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py]: SyncTokenPolicy sounds surprising in a sync file?

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py [hunk 4]
```diff
@@ -113,8 +129,40 @@ def _create_appconfig_pipeline(self, **kwargs):
...
+        if base_url.endswith('/'):
+            scope = base_url + '.default'
+        else:
+            scope = base_url + '/.default'
+
+        if policies is None:  # [] is a valid policy list
+            if hasattr(credential, "get_token"):
+                credential_policy = AsyncBearerTokenCredentialPolicy(credential, scope)
+            else:
+                raise TypeError("Unsupported credential: {}".format(credential))
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py]: "Please provide an instance from azure-identity or a class that implement the "get_token" protocol"

TLDR: Explain how to recover from the error

File: sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py [hunk 5]
```diff
@@ -27,12 +27,68 @@
...
+class AppConfigurationClientTest(AzureMgmtTestCase):
     def __init__(self, method_name):
-        super(AppConfigurationClientTest, self).__init__(
-            method_name, AzureAppConfigurationClient
+        super(AppConfigurationClientTest, self).__init__(method_name)
+        self.vcr.match_on = ["path", "method", "query"]
+        if self.is_playback():
+            connection_str = "Endpoint=https://fake_app_config.azconfig-test.io;Id=0-l4-s0:h5htBaY5Z1LwFz50bIQv;Secret=bgyvBgwsQIw0s8myrqJJI3nLrj81M/kzSgSuP4BBoVg="
+        else:
+            connection_str = os.getenv('APP_CONFIG_CONNECTION')
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py]: You can do something like this:
```python
self.connection_str = self.set_value_to_scrub("APP_CONFIG_CONNECTION", APP_CONFIG_CONNECTION)
```
That will read the env variable AZURE_APP_CONFIG_CONNECTION, and if not present will read APP_CONFIG_CONNECTION in the settings file, and raise if not present + it will scrub it automatically if the value is seen in the recordings.

Example:
https://github.com/Azure/azure-sdk-for-python/blob/8064bc5df9a3c086fc17572cb7a96d4b0fa5d1b4/sdk/keyvault/azure-keyvault-certificates/tests/certificates_preparer.py#L99

Comment (lmazuel) [reply sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py]: Also, can't you create the AppConfig using ARM? Like Storage and KeyVault are doing in their test? This ensure the AppConfig is deleted correctly.

File: sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py [hunk 6]
```diff
@@ -27,12 +27,68 @@
...
+                value=TEST_VALUE,
+                content_type=TEST_CONTENT_TYPE,
+                tags={"tag1": "tag1", "tag2": "tag2"},
+            )
         )
+        self.to_delete = [self.test_config_setting, self.test_config_setting_no_label]
+
+    def tearDown(self):
+        super(AppConfigurationClientTest, self).tearDown()
+        for item in self.to_delete:
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py]: Can't you have key specific per tests and delete the entire thing at the end?

File: sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client_aad.py [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+        try:
+            self.app_config_client.delete_configuration_setting(key=key, label=label)
+        except AzureError:
+            logging.debug(
+                "Error occurred removing configuration setting %s %s during unit test"
+                % (key, label)
+            )
+
+    # method: add_configuration_setting
+    @pytest.mark.live_tests
```
Comment (lmazuel) [inline sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client_aad.py]: Why this entire file is live only?

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py [hunk 8]
```diff
@@ -54,7 +55,16 @@ class AzureAppConfigurationClient:
     # pylint:disable=protected-access
 
     def __init__(self, base_url, credential, **kwargs):
-        # type: (str, AppConfigConnectionStringCredential, dict) -> None
+        # type: (str, any, dict) -> None
+        try:
+            if not base_url.lower().startswith('http'):
+                base_url = "https://" + base_url
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py]: I don't know where the inputs to this are coming from, is there a chance protocol-relative URLs e.g. `//some.url` could be passed in?

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py [hunk 9]
```diff
@@ -107,8 +123,37 @@ def _create_appconfig_pipeline(self, **kwargs):
...
+
+        return Pipeline(transport, policies)
+
+    def _create_appconfig_aad_pipeline(self, base_url, credential, **kwargs):
+        transport = kwargs.get('transport')
+        policies = kwargs.get('policies')
+        if base_url.endswith('/'):
+            scope = base_url + '.default'
+        else:
+            scope = base_url + '/.default'
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py]: maybe personal pref but I try to avoid conditionals for things like this:
```
scope = base_url.strip("/") + "/.default"
```
Also nit: I think Black will want double quotes on the literal

File: sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py (hunk 5 above)
Comment (xiangyan99) [reply sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py]: Will be done in https://github.com/Azure/azure-sdk-for-python/issues/9062

File: sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py (hunk 6 above)
Comment (xiangyan99) [reply sdk/appconfiguration/azure-appconfiguration/tests/test_azure_configuration_client.py]: Will be done in https://github.com/Azure/azure-sdk-for-python/issues/9062

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py (hunk 8 above)
Comment (xiangyan99) [reply sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py]: The valid urls are like
https://fake_app_config.azconfig-test.io
http://fake_app_config.azconfig-test.io   (though we will raise error in credential policy)
fake_app_config.azconfig-test.io   (we want to add https:// for them)

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py (hunk 3 above)
Comment (xiangyan99) [reply sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_azure_configuration_client_async.py]: The service team uses the term "Sync-Token" in the document. Hence we added policy "SyncTokenPolicy".
https://github.com/Azure/AppConfiguration/blob/master/docs/REST/consistency.md
